src/util/getThoughtCommands.ts

```diff
@@ -0,0 +1,79 @@
+import Commands, { commands } from '../@types/Commands'
+
+/**
+ * This is a utility for creating opening and closing markup tag.
+ */
+const createTag = (tags: string) => ({
+  open: `<${tags}>`,
+  close: `</${tags}>`,
+})
+
+const tags = {
+  bold: createTag('b'),
+  italic: createTag('i'),
+  underline: createTag('u'),
+  strikethrough: createTag('strike'),
+}
+
+/**
+ * This determines which commands (bold, italic, underline, strikethrough)
+ * apply to an entire thought string.
+ */
+const getThoughtCommands = (thought: string): Commands => {
```

Comment:
I like the idea here. The approach is a bit too heavy for what it's doing though. I recommend a simpler way to detect if an entire though is formatted:

Starts with `<b>`
&& Ends with `</b>`
&& Only has one instance of `<b>`

Reply:
I'm afraid to sacrifice correctness for performance if we have alternatives, so I tried benchmarking with the following test cases:

```
const loremIpsum =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.'

const testCases = [
  loremIpsum,
  `<b>${loremIpsum}</b>`,
  `<b><i><u><strike>${loremIpsum}</strike></u></i></b>`,
  new Array(100).fill(`<b><i><u><strike>${loremIpsum}</strike></u></i></b>`).join(' '),
]
```

Here's the average duration from 1000 rounds on each of those test cases:

```
0.169ms
0.188ms
0.257ms
17.047ms
```

This already isn't too bad considering that this path isn't called while the user is typing (only called when formatting/selection changes and the cursor isn't actually in a thought). However, we can do much better. The biggest bottleneck in `getThoughtCommands` is that we parse non-tag characters one a time. By tackling that low-hanging fruit in 2cde51700703d4e7abf124eb616fb1e016d55f79, I've brought the benchmarks down to these:

```
0.004ms
0.005ms
0.009ms
0.64ms
```

That seems completely acceptable to me. Would you agree @raineorshine? I'll be happy to simplify the logic as suggested if not, but my proposal is to continue as-is after the refactoring in 2cde51700703d4e7abf124eb616fb1e016d55f79.

Reply:
The problem is that there is a lot more complexity than seems necessary.

```ts
const queryCommandState(tag: string, s: string) => s.startsWith('<${tag}>') && s.endsWith('</${tag}>') && hasOnlyOne(s, '<${tag}>')
const commandState = {
  bold: queryCommandState('b', value),
  italic: queryCommandState('i', value),
  underline: queryCommandState('u', value),
  strikethrough: queryCommandState('strike', value)
}
```

I would need a good reason to adopt a procedural parser instead of the above, and as you pointed out, this function is not run on every keystroke, so it is probably not a performance bottleneck. Optimizing the performance at the expensive of simple, stateless solution would be a mistake.

Reply:
This would not correctly handle a thought which has multiple styles. Only the "bold" button would be active for the thought `"<b><i>example</i></b>"`, missing the "italic" button. I assume a thought with two styles such as that would be relatively common, and using the shortcut to add/remove italics would update the text but not be reflective in the active state of the toolbar button.

Reply:
Ah, I forget about that case. Thank you, that makes more sense now.

I'll take another look to see if there are any opportunities to simplify it without removing support for multiple formatting tags.